lua/nvim-tree/git/cli.lua

```diff
@@ -0,0 +1,136 @@
+    return
+  end
+  M._GIT_IN_PATH = true
+
+  if type(uv_spawn_opts) ~= "table" then
+    error("uv_spawn_opts table is required", 2)
+  end
+
+  if type(uv_spawn_opts.args) ~= "table" then
+    error("uv_spawn_opts.args table is required", 2)
+  end
+
+  if type(uv_spawn_opts.timeout) ~= "number" then
+    uv_spawn_opts.timeout = 10000 -- 8 seconds
+  end
+
+  local onErrorOrData_type = type(on_err_or_data)
+  if onErrorOrData_type ~= "function" then
+    error("on_err_or_data callback is required, got " .. onErrorOrData_type, 2)
+  end
+
+  local process
+  local process_exited = false
+
+  local stdin = uv.new_pipe()
+  local stdout = uv.new_pipe()
+  local stderr = uv.new_pipe()
+
+  uv_spawn_opts = vim.tbl_extend("keep", uv_spawn_opts, {
+    stdio = { stdin, stdout, stderr },
+  })
+
+  local stream_context = { failed = false }
+  local on_data = function(err, data)
+    local failed = stream_context.failed or false
+    if err then
+      if not failed then
+        stream_context.failed = err
+        on_err_or_data(err, data)
+      end
+      ---@diagnostic disable-next-line: ambiguity-1
+      M.notify.error("git" .. table.concat(uv_spawn_opts.args) .. " job has failed: " .. err or "")
+    else
+      on_err_or_data(nil, data)
+    end
+  end
+
+  -- default exit handler; just reporting
+  --- @diagnostic disable-next-line: redefined-local
+  -- luacheck: ignore shadowing an upvalue
+  local _on_exit = function(process_exit_code, signal)
+    process_exited = true
+    if process_exit_code ~= 0 then
+      log.line("git", "git " .. table.concat(uv_spawn_opts.args) .. " job has failed with %s exit code", process_exit_code)
+    end
+    -- free FIFOs
+    -- as uv library requires for handles to be closed upon exit
+    -- ref: https://docs.libuv.org/en/v1.x/guide/processes.html#spawning-child-processes
+    stdin:shutdown(function()
+      process:close()
+    end)
+    stdout:shutdown()
+    stderr:shutdown()
+
+    if on_exit then
+      on_exit(process_exit_code, signal)
+    end
+  end
+
+  M.log.line("git", "running job with timeout %dms", uv_spawn_opts.timeout)
+  M.log.line("git", "git %s", table.concat(uv_spawn_opts.args, " "))
+
+  process = uv.spawn("git", uv_spawn_opts, _on_exit)
+  uv.read_start(stdout, vim.schedule_wrap(on_data))
+  uv.read_start(stderr, vim.schedule_wrap(on_data))
+
+  -- If git command takes too long to run, kill the process
+  -- the option .timeout have to be explicitly set to activate this featureg;
+  vim.defer_fn(function()
+    if process_exited or stream_context.failed or (process and process:is_closing()) then
+      return
+    end
+    local process_kill_failed = process:kill()
+    if process_kill_failed == 0 then
+      if on_timeout then
+        on_timeout()
+      end
+    end
+  end, uv_spawn_opts.timeout)
+end
+
+-- note: keep this in sync with M.cli
+
+M.git = M.cli
+-- This is a trick to workaround lua_lsp_server limitation on metatable
+-- function type declarations.
```

Comment:
How and why does this exceed limitations?

We should resolve the root cause rather than working around limitations.

Reply:
@alex-courtis This is limitation of the lua-language-server (LLS). LLS no `@function` tag to describe var as function + params. The workaround alows to get the same completion signature for both `M.git` and `M.cli` without extra LuaDOC declarations.

Reply:
Why do we need to force type to coerce the signatures?

Can you not just use unique names i.e. do not overload `.git`?

Unnecessary fragile code to work around language server limitations is not acceptable.